File: AndroidLearnProject/basemodule/src/main/java/com/android/learn/base/utils/SPUtils.java
package com.android.learn.base.utils;
 
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Map;
 
import android.content.Context;
import android.content.SharedPreferences;

import com.android.learn.base.application.CustomApplication;

public class SPUtils
{
	/**
	 * 保存在手机里面的文件名
	 */
	public static final String FILE_NAME = "share_data";
 

	public static void put(String key, Object object)
	{
 
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
		SharedPreferences.Editor editor = sp.edit();
 
		if (object instanceof String)
		{
			editor.putString(key, (String) object);
		} else if (object instanceof Integer)
		{
			editor.putInt(key, (Integer) object);
		} else if (object instanceof Boolean)
		{
			editor.putBoolean(key, (Boolean) object);
		} else if (object instanceof Float)
		{
			editor.putFloat(key, (Float) object);
		} else if (object instanceof Long)
		{
			editor.putLong(key, (Long) object);
		} else
		{
			editor.putString(key, object.toString());
		}
 
		SharedPreferencesCompat.apply(editor);
	}
 

	public static Object get(String key, Object defaultObject)
	{
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
 
		if (defaultObject instanceof String)
		{
			return sp.getString(key, (String) defaultObject);
		} else if (defaultObject instanceof Integer)
		{
			return sp.getInt(key, (Integer) defaultObject);
		} else if (defaultObject instanceof Boolean)
		{
			return sp.getBoolean(key, (Boolean) defaultObject);
		} else if (defaultObject instanceof Float)
		{
			return sp.getFloat(key, (Float) defaultObject);
		} else if (defaultObject instanceof Long)
		{
			return sp.getLong(key, (Long) defaultObject);
		}
 
		return null;
	}
 

	public static void remove( String key)
	{
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
		SharedPreferences.Editor editor = sp.edit();
		editor.remove(key);
		SharedPreferencesCompat.apply(editor);
	}
 
	/**
	 * 清除所有数据
	 */
	public static void clear()
	{
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
		SharedPreferences.Editor editor = sp.edit();
		editor.clear();
		SharedPreferencesCompat.apply(editor);
	}
 
	/**
	 * 查询某个key是否已经存在
	 */
	public static boolean contains(String key)
	{
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
		return sp.contains(key);
	}
 
	/**
	 * 返回所有的键值对
	 */
	public static Map<String, ?> getAll()
	{
		SharedPreferences sp = CustomApplication.context.getSharedPreferences(FILE_NAME,
				Context.MODE_PRIVATE);
		return sp.getAll();
	}
 
	/**
	 * 创建一个解决SharedPreferencesCompat.apply方法的一个兼容类
	 * 
	 * @author zhy
	 * 
	 */
	private static class SharedPreferencesCompat
	{
		private static final Method sApplyMethod = findApplyMethod();
 
		/**
		 * 反射查找apply的方法
		 * 
		 * @return
		 */
		@SuppressWarnings({ "unchecked", "rawtypes" })
		private static Method findApplyMethod()
		{
			try
			{
				Class clz = SharedPreferences.Editor.class;
				return clz.getMethod("apply");
			} catch (NoSuchMethodException e)
			{
			}
 
			return null;
		}
 
		/**
		 * 如果找到则使用apply执行，否则使用commit
		 * 
		 * @param editor
		 */
		public static void apply(SharedPreferences.Editor editor)
		{
			try
			{
				if (sApplyMethod != null)
				{
					sApplyMethod.invoke(editor);
					return;
				}
			} catch (IllegalArgumentException e)
			{
			} catch (IllegalAccessException e)
			{
			} catch (InvocationTargetException e)
			{
			}
			editor.commit();
		}
	}
 
}

File: basemodule/src/main/java/com/android/learn/base/utils/SPUtils.kt
package com.android.learn.base.utils

import android.content.Context
import android.content.SharedPreferences


object SPUtils {

    /**
     * 保存在手机里面的文件名
     */
    private val FILE_NAME = "SP_FILE"


    /**
     * 保存数据的方法，我们需要拿到保存数据的具体类型，然后根据类型调用不同的保存方法
     * @param context
     * @param key
     * @param object
     */
    fun setParam(context: Context, key: String, `object`: Any) {

        val type = `object`.javaClass.simpleName
        val sp = context.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)
        val editor = sp.edit()

        if ("String" == type) {
            editor.putString(key, `object` as String)
        } else if ("Integer" == type) {
            editor.putInt(key, `object` as Int)
        } else if ("Boolean" == type) {
            editor.putBoolean(key, `object` as Boolean)
        } else if ("Float" == type) {
            editor.putFloat(key, `object` as Float)
        } else if ("Long" == type) {
            editor.putLong(key, `object` as Long)
        }

        editor.commit()
    }


    /**
     * 得到保存数据的方法，我们根据默认值得到保存的数据的具体类型，然后调用相对于的方法获取值
     * @param context
     * @param key
     * @param defaultObject
     * @return
     */
    fun getParam(context: Context, key: String, defaultObject: Any): Any? {
        val type = defaultObject.javaClass.simpleName
        val sp = context.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)

        if ("String" == type) {
            return sp.getString(key, defaultObject as String)
        } else if ("Integer" == type) {
            return sp.getInt(key, defaultObject as Int)
        } else if ("Boolean" == type) {
            return sp.getBoolean(key, defaultObject as Boolean)
        } else if ("Float" == type) {
            return sp.getFloat(key, defaultObject as Float)
        } else if ("Long" == type) {
            return sp.getLong(key, defaultObject as Long)
        }

        return null
    }

    /**
     * 清除所有数据
     * @param context
     */
    fun clearAll(context: Context) {
        val sp = context.getSharedPreferences(FILE_NAME,
                Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.clear().commit()
    }

    /**
     * 清除指定数据
     * @param context
     */
    fun clear(context: Context, key: String) {
        val sp = context.getSharedPreferences(FILE_NAME,
                Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.remove(key)
        editor.commit()
    }

}
